vmm/src/device_manager.rs

```diff
@@ -1629,3 +1637,24 @@ impl Drop for DeviceManager {
         }
     }
 }
+
+impl Pausable for DeviceManager {
+    fn pause(&mut self) -> result::Result<(), MigratableError> {
+        for dev in &self.migratable_devices {
```

Comment:
What I was mentioning in my previous comment is that we could maybe iterate over all the devices found on both io_bus and mmio_bus, as they actually own these devices. This way we don't need the device manager to own the devices, but just get a reference at the right time. In this case, the right time is to retrieve the reference and call the `pause()` method on it.

Reply:
Well, the `CpuManager` is on the io_bus so you'd need to be careful.

Reply:
> What I was mentioning in my previous comment is that we could maybe iterate over all the devices found on both io_bus and mmio_bus, as they actually own these devices. This way we don't need the device manager to own the devices, but just get a reference at the right time. In this case, the right time is to retrieve the reference and call the `pause()` method on it.

That's also assuming that each device on these bus implement the `Migratable` trait, which is not guaranteed.

Reply:
>That's also assuming that each device on these bus implement the Migratable trait, which is not guaranteed.

Well if they don't, the getter would simply return `None` instead of a reference to the actual device.

Because almost every single device will need to be `Migratable`, we're going to end up locking every device, which does not sound great for the whole codebase.
Can you run a quick `perf stat` with and without this PR? I'd like to make sure the time spent locking is not increasing excessively.

Reply:
Running perf on a quick test that exercises the virtio console device, then looking for `Mutex` symbols:

### Master branch

```
   0.13%  virtio_console  cloud-hypervisor  [.] std::sync::mutex::Mutex<T>::lock
   0.10%  virtio_console  cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::lock
   0.07%  virtio_console  cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::deref::DerefMut>::deref_mut
   0.04%  virtio_console  cloud-hypervisor  [.] std::sync::mutex::MutexGuard$LT$T$GT$::new::_$u7b$$u7b$closure$u7d$$u7d$::h263bbc7fea6c1311
   0.04%  virtio_console  cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::unlock
   0.03%  virtio_console  cloud-hypervisor  [.] std::sync::mutex::MutexGuard<T>::new
   0.02%  virtio_console  cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::drop::Drop>::drop
   0.01%  virtio_console  cloud-hypervisor  [.] std::sys_common::mutex::Mutex::raw_unlock
   0.01%  virtio_console  cloud-hypervisor  [.] std::sys_common::mutex::Mutex::raw_lock
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::lock
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::unlock
   0.00%  virtio_blk      cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::deref::Deref>::deref
   0.00%  vcpu0           cloud-hypervisor  [.] std::sync::mutex::Mutex<T>::lock
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sync::mutex::Mutex<T>::lock
   0.00%  vcpu0           cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::deref::DerefMut>::deref_mut
   0.00%  vcpu0           cloud-hypervisor  [.] std::sync::mutex::MutexGuard$LT$T$GT$::new::_$u7b$$u7b$closure$u7d$$u7d$::habad5377374a1f76
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sync::mutex::MutexGuard$LT$T$GT$::new::_$u7b$$u7b$closure$u7d$$u7d$::h4fc865ef5c2cd668
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sys_common::mutex::Mutex::raw_lock
   0.00%  virtio_blk      cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::drop::Drop>::drop
```

### Pausable branch
```
   0.10%  virtio_console  cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::deref::DerefMut>::deref_mut
   0.08%  virtio_console  cloud-hypervisor  [.] std::sync::mutex::Mutex<T>::lock
   0.08%  virtio_console  cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::lock
   0.07%  virtio_console  cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::drop::Drop>::drop
   0.06%  virtio_console  cloud-hypervisor  [.] std::sync::mutex::MutexGuard<T>::new
   0.03%  virtio_console  cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::unlock
   0.02%  virtio_console  cloud-hypervisor  [.] std::sync::mutex::MutexGuard$LT$T$GT$::new::_$u7b$$u7b$closure$u7d$$u7d$::h945747274e138860
   0.02%  virtio_console  cloud-hypervisor  [.] std::sys_common::mutex::Mutex::raw_unlock
   0.02%  virtio_console  cloud-hypervisor  [.] std::sys_common::mutex::Mutex::raw_lock
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sync::mutex::Mutex<T>::lock
   0.00%  virtio_blk      cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::deref::Deref>::deref
   0.00%  vcpu0           cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::unlock
   0.00%  vcpu0           cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::drop::Drop>::drop
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::unlock
   0.00%  virtio_console  cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::deref::Deref>::deref
   0.00%  virtio_blk      cloud-hypervisor  [.] std::sys::unix::mutex::Mutex::lock
   0.00%  virtio_blk      cloud-hypervisor  [.] <std::sync::mutex::MutexGuard<T> as core::ops::drop::Drop>::drop
```

Reply:
Well this looks pretty good :)
I think we can go ahead with this PR then.